Remove at::Optional

Yangqing opened the pull request:
Summary: TSIA - this is an ongoing diff to fully move to c10 namespace.

Differential Revision: D10494123

ezyang commented:
It seems to me that, order of operations wise, we should land #12950 first, and that will eliminate most of the codemodding in this diff. Otherwise, we have to whiplash it to `c10::optional`, and then whiplash it back shortly afterwards. That makes people working on diffs that touch these lines grumpy, because it makes extra merge conflicts. Let's not make people grumpy.

Commit:
Remove at::Optional (#12958)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/12958

TSIA - this is an ongoing diff to fully move to c10 namespace.

Differential Revision: D10494123

fbshipit-source-id: 67e5b61a922b663f96a321b48ccc9cd24dcde8be
